Factor out information unit formatting

awemany opened the pull request:
There's a couple places where a floating point amount of bytes is
formatted  with units attached (KB, MB, ...).

This factors all of them out and puts them into a single function
static string formatInfoUnit(double).

It also fixes a couple potential array-out-of-bounds-access errors,
as none of the upper bounds had been checked before.

awemany committed:
Factor out information unit formatting

There's a couple places where a floating point amount of bytes is
formatted  with units attached (KB, MB, ...).

This factors all of them out and puts them into a single function
string formatInfoUnit(double).

It also fixes a couple potential array-out-of-bounds-access errors,
as none of the upper bounds had been checked before, and adds docs
and a unit test.

kyuupichan commented on src/thinblock.cpp:
```diff
@@ -28,6 +28,22 @@ extern CCriticalSection cs_thinblockstats;
+{
+    static const char *units[] = {"B", "KB", "MB", "GB", "TB", "PB", "EB", "ZB", "YB"};
+    int i = 0;
+    while (value > 1000.0 && i < 7)
```
7 should not be hard-coded.  0...6 is probably enough anyway, no need for the ZB and YB

awemany commented:
@kyuupichan : Removed ZB and YB, calculated upper bound from array size.
@ftrader : Added unit test for formatInfoUnit(..).

ftrader commented:
I'm not a great fan of retaining the decimal point with only trailing zeroes (.00), they convey no info to me, but that's a niggle and I'm fine with the changes.

gandrewstone commented on src/thinblock.cpp:
```diff
@@ -28,6 +28,26 @@ extern CCriticalSection cs_thinblockstats;
+    static const char *units[] = {"B", "KB", "MB", "GB", "TB", "PB", "EB"};
+
+    if (value < 0)
+        value = 0.0;
```
I'm concerned about mis-displaying negative numbers, even if we don't have them right now.  I think you can make it work by removing the above test and using while ((value < -1000.0 || value > 1000.0) .  Can you try it and extend the tests to negatives?

awemany replied:
Ok so you want -10000 be displayed at -10.00KB for example?

Sure I can do that. Now I just wonder: does it make sense to move the whole function somewhere else, like generic util.cpp, or unlimited.cpp?

Or is it fine where it is?

gandrewstone commented on src/thinblock.h:
```diff
@@ -18,6 +18,8 @@
 class CDataStream;
 class CNode;
 
+extern std::string formatInfoUnit(double value);
```
add a docstring comment describing the operation of this function.  See https://github.com/BitcoinUnlimited/BitcoinUnlimited/blob/release/doc/developer-notes.md#doxygen-comments

awemany commented:
@gandrewstone: Addressed your points, hopefully to your liking. Also added the test to the list of properly formatted files.

awemany commented:
The failure is in parallel.py on just one build - all others are fine, if I interpret the travis stats correctly.

I fail to see how this is related to my code. Might this be a spurious failure (race?) in the parallel validation code? @ptschip?

I updated my code to add formatting for negative values and added comments - I can't see how that might have created this failure. I *also* updated to the most current `dev` branch HEAD.

gandrewstone commented:
looks great, I'll commit as soon as tests pass
